# Matrizes e vetores

## License

All content can be freely used and adapted under the terms of the 
[Creative Commons Attribution 4.0 International License](http://creativecommons.org/licenses/by/4.0/).

![Creative Commons License](https://i.creativecommons.org/l/by/4.0/88x31.png)

## Representação de uma matriz

Antes de podermos manipular matrizes e vetores no computador, precisamos de um jeito de armazená-los em variáveis.
Para vetores, o candidato natural são **listas**.

$$
\mathbf{v} = \begin{bmatrix}1 \\ 2 \\ 3\end{bmatrix}
$$

O vetor acima poderia ser representado em código como:



In [370]:
v = [1, 2, 3]
print(v)

[1, 2, 3]


Uma matriz pode ser vista como um conjunto de vetores, ou um vetor de vetores. Cada vetor seria equivalente a uma linha da matriz:

$$
\mathbf{A} =
\begin{bmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9 \\
\end{bmatrix} =
\begin{bmatrix}
[1 & 2 & 3] \\
[4 & 5 & 6] \\
[7 & 8 & 9] \\
\end{bmatrix}
$$

Logo, um jeito de representar uma matriz em Python é através de uma **lista de listas**:

In [371]:
A = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
print(A)

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]


O Python permite quebrar a linha quando o comando está entre `[` ou `(`:

In [372]:
A = [[1, 2, 3],
     [4, 5, 6],
     [7, 8, 9]]
print(A)

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]


*Note que o print acima imprime a matriz toda em uma única linha*.

A matriz `A` é uma lista como qualquer outra. A única diferença é que **os elementos da lista são outras listas**. O elemnto `[0]` da lista `A` será a lista correspondente a primeira linha:

In [373]:
print(A[0])

[1, 2, 3]


Como `A[0]` é uma lista, podemos pegar os elementos dessa lista da mesma forma:

In [374]:
print(A[0][0])
print(A[1][2])

1
6


### Exemplo

Queremos imprimir cada elemento da matriz `A` da seguinte forma:

    1  2  3  
    4  5  6  
    7  8  9

Podemos utilizar o `for` para pegar cada linha da matriz. Para cada linha da matriz, queremos imprimir o elemento. Quando terminarmos de imprimir a linha, vamos pular uma linha.

In [375]:
for i in range(3): # Anda sobre as linhas
    for j in range(3): # Anda sobre as colunas
        print(A[i][j], ' ', end='') # end='' faz com que print não pule uma linha
    print() # Imprime nada e pula uma linha 

1  2  3  
4  5  6  
7  8  9  


## Somando matrizes

A soma de duas matrizes produz uma matriz que terá a soma dos elementos correspondentes:

$$
\begin{bmatrix}
a & b & c \\
d & e & f \\
g & h & i \\
\end{bmatrix} +
\begin{bmatrix}
j & l & m \\
n & o & p \\
q & r & s \\
\end{bmatrix} =
\begin{bmatrix}
a + j & b + l & c + m \\
d + n & e + o & f + p \\
g + q & h + r & i + s \\
\end{bmatrix}
$$

De forma genérica, o $j$-ésimo elemento da $i$-ésima linha de uma matriz $\mathbf{A}$ é $A_{ij}$.
Se temos duas matrizes $\mathbf{A}$ e $\mathbf{B}$, a soma pode ser escrita como:

$$
C_{ij} = A_{ij} + B_{ij}
$$

## Tarefa

Some as matrizes `A` e `B` definidas abaixo e guarde o resultado em uma matriz `C`. Imprima a matriz `C`.

**Dicas**:

* Você pode criar a matriz `C` antes ou durante a soma. Lembre-se que a matriz é uma lista e listas possuem o método `append`.

In [376]:
A = [[1, 2, 3],
     [4, 5, 6],
     [7, 8, 9]]
B = [[3, 4, 5],
     [6, 7, 8],
     [9, 10, 11]]
nlin_a = 3
nlin_b = 3
ncol_a = 3
ncol_b = 3


In [377]:
def somar(m1, m2): #definimos uma função para efetuar a soma das matrizes
    matriz_soma = []
    # Supondo que as duas matrizes possuem o mesmo tamanho
    quant_linhas = len(m1) # Conta quantas linhas existem
    quant_colunas = len(m1[0]) # Conta quantos elementos têm em uma linha
    for i in range(quant_linhas):
        # Cria uma nova linha na matriz_soma
        matriz_soma.append([])
        for j in range(quant_colunas):
            # Somando os elementos que possuem o mesmo índice (por isso o for de j está dentro do for de i)
            soma = m1[i][j] + m2[i][j]
            matriz_soma[i].append(soma)
    return matriz_soma

In [378]:
C = somar (A, B) #aplicamos a função nas matrizes que queriamos somar

In [379]:
for i in range(3):
    for j in range(3):
        print(C[i][j], ' ', end='')
    print()

4  6  8  
10  12  14  
16  18  20  


### Resultado esperado

Seu código deve imprimir exatamente:

    4  6  8  
    10  12  14  
    16  18  20  

## Multiplicando uma matriz por um vetor

A multiplicação de uma matriz por um vetor é

$$
\begin{bmatrix}
a & b  \\
c & d  \\
\end{bmatrix} 
\begin{bmatrix}
e \\
f \\
\end{bmatrix} =
\begin{bmatrix}
ae + bf \\
ce + df \\
\end{bmatrix}
$$

Sendo o vetor $\mathbf{u} = \mathbf{A}\mathbf{v}$, cada elemento $i$ de $\mathbf{u}$ é 


$$
u_i = \sum\limits_{k=1}^{N} A_{ik}v_k
$$

## Tarefa

Faça a multiplicação da matriz pelo vetor definidos abaixo. Guarde o resultado em uma lista.

**Dicas**

In [380]:
A = [[1, 2, 3, 4],
     [4, 5, 6, 7],
     [7, 8, 9, 10]]
v = [12, 13, 14, 15]
nlin = 3
ncol = 4

In [381]:
def mult_vm(m1,v1):    #criamos uma função para fazer a multiplicação da matriz pelo vetor
    n_linhas = len(m1) # Conta quantas linhas existem na matriz
    resp = []          #criamos a lista, onde vai entrar nossa resposta
    for i in range(n_linhas): #usamos for junto a range para fazer a operação pra cada elemento da linha
        mult = 0        #consideramos que nossa operação começara do zero
        n_colunas = len(m1[i]) 
        for j in range(n_colunas):
        #usamos esse for dentro do outro para posicionar o elemento de acordo com a lina e coluna em que se encontra
            mult = mult + v1[j] * m1[i][j] 
            #para que houvesse a soma dos elementos da mesma linha multiplicados pelo vetor criamos essa operação
        resp.append(mult) #guardamos o resultado na lista
    return resp

In [382]:
C = mult_vm(A,v)
print (C)


[140, 302, 464]


### Resultado esperado

O seu código deve imprimir exatamente:

    [140, 302, 464]

## Multiplicação de matrizes

A multiplicação de matrizes é feita de forma diferente da soma. É mais fácil mostrar do que explicar:

$$
\begin{bmatrix}
a & b  \\
c & d  \\
\end{bmatrix} 
\begin{bmatrix}
e & f \\
g & h  \\
\end{bmatrix} =
\begin{bmatrix}
ae + bg & af + bh  \\
ce + dg & cf + dh  \\
\end{bmatrix}
$$

Sendo $\mathbf{C} = \mathbf{A}\mathbf{B}$, cada elemento $ij$ de $\mathbf{C}$ é 


$$
C_{ij} = \sum\limits_{k=1}^{N} A_{ik}B_{kj}
$$

## Tarefa

Faça a multiplicação das duas matrizes definidas abaixo. Guarde o resultado em uma matriz (lista de listas).

In [383]:
A = [[1, 2, 3, 4],
     [4, 5, 6, 7],
     [7, 8, 9, 10]]
B = [[3, 4, 5],
     [6, 7, 8],
     [9, 10, 11],
     [12, 13, 14]]
nlin_a = 3
nlin_b = 4
ncol_a = 4
ncol_b = 3

In [384]:
def mult_mm(m1,m2):    #definimos uma função para multiplicação de matrizes
    linhas=len(m1)     #contamos o numero de linhas na matriz m1
    colunas=len(m2[0]) #contamos o numero de colunas na matriz m2
    C=[]               #criamos uma lista para colocar nossa resposta
    for i in range(linhas): #usamos for com range para contar os elementos contidos na linha da matriz m1
        C.append([])
        for p in range(colunas): 
        #usamos este for com range dentro do anterior para identificar a linha e coluna de cada elemento da matriz m1
            val = 0 #consideramos que nossa operação vai começar do zero
            for k in range (len(m2)): 
#considerando que o numero de linhas deve ser igual ao da matriz m1 adicionamos esse for para contar o numero de colunas da matriz m2. 
                    val = val + m1[i][k]*m2[k][p] #usamos ets operação levando em conta que cada elemento c[i][j] da matriz resposta é a soma do elemtos  a[i][k] da mesma linha da matriz m1 com os elemtos b[k][j] da mesma coluna na matriz m2. 
            C[i].append(val)
    return C 

In [385]:
C = mult_mm(A,B)
for i in range(3):
    for j in range(3):
        print(C[i][j], ' ', end='')
    print()

90  100  110  
180  202  224  
270  304  338  


### Resultado esperado

O seu código deve imprimir exatamente:

    90  100  110  
    180  202  224  
    270  304  338  



## Tarefa Bônus

Calcule o produto $\mathbf{A}^T\mathbf{A}$ para a matriz $A$ definida abaixo. $\mathbf{A}^T$ é a matriz transposta de $A$:

$$
\mathbf{A} = 
\begin{bmatrix}
a & b & c \\
d & e & f \\
\end{bmatrix}
$$

$$
\mathbf{A}^T = 
\begin{bmatrix}
a & d \\
b & e \\
c & f \\
\end{bmatrix}
$$

Ou seja, as linhas de $\mathbf{A}$ são as colunas de $\mathbf{A}^T$.


In [386]:
A = [[1, 2, 3, 4],
     [4, 5, 6, 7],
     [7, 8, 9, 10]]
nlin = 3
ncol = 4 

In [387]:
def mat_trans (M): #criamos uma função para fazer a matriz transposta
    trans = []  #criamos uma lista para receber nossa resposta
    for i in range(len(M[0])): #usamos este for com range para contar as colunas e usa-las para fazer a transposição
        aux =[] #criamos uma lista auxiliar para ajudar na operação
        for k in M: #usamos este for para utlizar cada elemento dentro da matriz
            aux.append(k[i]) 
    #adicionou os elementos de acordo com as linhas na lista auxiiar (correspondentes a coluna da matriz original)
        trans.append(aux) #adicionamos a lista na matriz para obter a posição na linha e coluna pra cada elemento.
   
    return trans

In [388]:
C = mat_trans(A)
for i in range(4): # Anda sobre as linhas
    for j in range(3): # Anda sobre as colunas
        print(C[i][j], ' ', end='') # end='' faz com que print não pule uma linha
    print() # Imprime nada e pula uma linha 

1  4  7  
2  5  8  
3  6  9  
4  7  10  


In [394]:
D = mult_mm(C,A) #aplicamos a função de multiplicação de matrizes
for i in range(4):
    for j in range(4):
        print(D[i][j], ' ', end='')
    print()

66  78  90  102  
78  93  108  123  
90  108  126  144  
102  123  144  165  


### Resultado esperado

O seu código deve imprimir exatamente:

    66  78  90  102  
    78  93  108  123  
    90  108  126  144  
    102  123  144  165  
